File: js/session-quality.js
// Fallback for older snapshots/API responses; kept in parity with session_quality.py.
// Policy v3: yield counter is unreliable -> segment on >30 min gap only, and
// reconstruct volume by integrating flow (flow_lpm * minutes / 1000) per session.
(function (root) {
  function build(series) {
    const times = series.times || [], water = series.water_ft || [],
      flow = series.flow_lpm || [], yield_ = series.yield_kl || [];
    if (!times.length) return [];
    const boundaries = [0];
    for (let i = 1; i < times.length; i++) {
      if (new Date(times[i]) - new Date(times[i - 1]) > 1800000) boundaries.push(i);
    }
    boundaries.push(times.length);
    const result = [];
    for (let b = 0; b < boundaries.length - 1; b++) {
      const a = boundaries[b], end = boundaries[b + 1], e = end - 1, n = end - a;
      const missingWater = water.slice(a, end).some(v => v == null);
      const missingYield = yield_.slice(a, end).some(v => v == null);
      const missingFlow = flow.slice(a, end).some(v => v == null);
      let maxStep = 0, jumps = 0;
      for (let i = a + 1; i < end; i++) {
        if (water[i] == null || water[i - 1] == null) continue;
        const step = Math.abs(water[i] - water[i - 1]);
        maxStep = Math.max(maxStep, step);
        if (step > 20) jumps++;
      }
      const drawdown = water[e] != null && water[a] != null ? water[e] - water[a] : null;
      let volume = 0;
      for (let i = a + 1; i < end; i++) {
        if (flow[i - 1] != null) {
          const minutes = (new Date(times[i]) - new Date(times[i - 1])) / 60000;
          if (minutes >= 0) volume += flow[i - 1] * minutes / 1000;
        }
      }
      const reasons = [];
      if (n < 3) reasons.push('too_few_samples');
      if (volume <= 0) reasons.push('no_volume');
      if (jumps) reasons.push('sensor_relock_jump');
      if (drawdown != null && drawdown < 0) reasons.push('net_level_rise');
      if (missingWater) reasons.push('missing_water_level');
      if (missingYield) reasons.push('missing_yield');
      const excluded = n < 3 || volume <= 0 || drawdown == null;
      const status = excluded ? 'excluded' : reasons.length ? 'flagged' : 'ok';
      const volumeOk = n >= 2 && !missingFlow && volume > 0;
      const drawdownOk = n >= 3 && !missingWater && !jumps && drawdown != null && drawdown > 0;
      const round = v => v == null ? null : Math.round((v + Number.EPSILON) * 100) / 100;
      result.push({ start: a, stop: e, n, drawdown_ft: round(drawdown), pumped_kl: round(volume),
        max_step_ft: round(maxStep), jump_count: jumps, status, usable: status === 'ok', reasons,
        eligible_volume: volumeOk, eligible_drawdown: drawdownOk,
        eligible_specific_capacity: volumeOk && drawdownOk });
    }
    return result;
  }
  const api = { build, version: 3 };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else root.SessionQuality = api;
})(typeof window !== 'undefined' ? window : globalThis);

File: backend/session_quality.py
"""KH flags with a separate, versioned dashboard inclusion policy.

No observations are removed or corrected. 'excluded' means excluded from the
default session analysis; metric eligibility remains independent of that label.
"""
POLICY_VERSION = 2


def build_sessions(times, water, flow, yield_):
    if not times:
        return []
    boundaries = [0]
    for i in range(1, len(times)):
        reset = (yield_[i] is not None and yield_[i - 1] is not None
                 and yield_[i] < yield_[i - 1])
        if reset or (times[i] - times[i - 1]).total_seconds() > 1800:
            boundaries.append(i)
    boundaries.append(len(times))
    result = []
    for a, end in zip(boundaries, boundaries[1:]):
        e = end - 1
        n = end - a
        missing_water = any(v is None for v in water[a:end])
        missing_yield = any(v is None for v in yield_[a:end])
        steps = [abs(water[i] - water[i - 1]) for i in range(a + 1, end)
                 if water[i] is not None and water[i - 1] is not None]
        jumps = sum(v > 20 for v in steps)
        drawdown = water[e] - water[a] if water[e] is not None and water[a] is not None else None
        volume = yield_[e] - yield_[a] if yield_[e] is not None and yield_[a] is not None else None
        reasons = []
        if n < 3:
            reasons.append('too_few_samples')
        if volume is not None and volume <= 0:
            reasons.append('no_volume')
        if jumps:
            reasons.append('sensor_relock_jump')
        if drawdown is not None and drawdown < 0:
            reasons.append('net_level_rise')
        # Missing measurements are additional dashboard checks, not KH rules.
        if missing_water:
            reasons.append('missing_water_level')
        if missing_yield:
            reasons.append('missing_yield')
        excluded = n < 3 or volume is None or volume <= 0 or drawdown is None
        status = 'excluded' if excluded else ('flagged' if reasons else 'ok')
        volume_ok = n >= 2 and not missing_yield and volume is not None and volume > 0
        drawdown_ok = n >= 3 and not missing_water and not jumps and drawdown is not None and drawdown > 0
        result.append({
            'start': a, 'stop': e, 'n': n,
            'drawdown_ft': round(drawdown, 2) if drawdown is not None else None,
            'pumped_kl': round(volume, 2) if volume is not None else None,
            'max_step_ft': round(max(steps, default=0), 2),
            'jump_count': jumps, 'status': status,
            'usable': status == 'ok', 'reasons': reasons,
            'eligible_volume': volume_ok, 'eligible_drawdown': drawdown_ok,
            'eligible_specific_capacity': volume_ok and drawdown_ok,
        })
    return result


def summarize(sessions):
    counts = {'total': len(sessions), 'ok': 0, 'flagged': 0, 'excluded': 0,
              'eligible_volume': 0, 'eligible_drawdown': 0,
              'eligible_specific_capacity': 0, 'jump_events': 0, 'by_reason': {}}
    for session in sessions:
        counts[session['status']] += 1
        counts['jump_events'] += session['jump_count']
        for key in ('eligible_volume', 'eligible_drawdown', 'eligible_specific_capacity'):
            counts[key] += int(session[key])
        for reason in session['reasons']:
            counts['by_reason'][reason] = counts['by_reason'].get(reason, 0) + 1
    return counts
